End to end (e2e) tests implemented along with supporting code and pro…

janybravo opened the pull request:
Added e2e test configuration, updated code for first test and simplified BE proxy configuration.

dblenkus commented on .env:
```diff
@@ -1,4 +1,5 @@
+REACT_APP_REST_AUTH_URL=http://localhost:3000/rest-auth
 REACT_APP_SENTRY_URL=https://e0a554a693f641a6923a4240a2a64dfb@o74924.ingest.sentry.io/5265645
 REACT_APP_WEBSOCKET_URL=wss://qa2.genialis.com/ws/
+SKIP_PREFLIGHT_CHECK=true
```
Each file should end with a new line. Also, add a comment why the variable is needed.

janybravo replied:
not actually needed, removed

dblenkus commented on Jenkinsfile:
```diff
@@ -30,7 +30,7 @@ throttle(["dictyexpress_js"]) {
                 }
             }
-
+            
```
Remove trailing spaces. It is the best if you configure the editor to do that on each save.

zigasinko replied:
Run `npm run check` to see which files need styling corrections (there are a few: changelog, cypress.json, ypress\tsconfig.json, reame, tsconfig.json).

janybravo replied:
@dblenkus its prettier that adds trailing spaces, ill leave as is.

dblenkus commented on cypress/tsconfig.json:
```diff
@@ -0,0 +1,10 @@
+  "compilerOptions": {
+    "noEmit": false,
+  }
+}
```
Add a new line.

janybravo replied:
added. json files are excluded from eslint as its hard to configure typescript-eslint/parser to check them without errors. Will tckle that anoter time, perhaps for expressions project

dblenkus commented on Jenkinsfile:
```diff
@@ -42,13 +42,25 @@ throttle(["dictyexpress_js"]) {
+                "E2E tests": {
+                    stage("Tests") {
+                        environment {
+                            USERNAME = credentials('jenkins-expressions-login-username')
```
`USERNAME` and `PASSWORD` variables are set, but then the code uses `LOGIN_USERNAME` and `LOGIN_PASSWORD`.

janybravo replied:
simplified var namings

dblenkus commented on package.json:
```diff
@@ -38,17 +38,19 @@
+        "test": "react-scripts test --env=jest-environment-jsdom-sixteen",
+        "cy": "cross-env CYPRESS_LOGIN_USERNAME=${USERNAME:-ci-genialis} CYPRESS_LOGIN_PASSWORD=${PASS:-wrong123} cypress",
+        "cy-run": "cypress run",
+        "test:e2e": "cross-env CYPRESS_LOGIN_USERNAME=${USERNAME:-ci-genialis} CYPRESS_LOGIN_PASSWORD=${PASS:-wrong123} start-server-and-test start 3000 cy-run",
```
Why do you rename `USERNAME` into `CYPRESS_LOGIN_USERNAME`? It is better to have explicite names from the beginning.

dblenkus commented on cypress/support/index.js:
```diff
@@ -0,0 +1,15 @@
+    cy.get('[data-testid=open-login-modal]').click();
+    // For debugging purposes
+    // expect(Cypress.env('LOGIN_USERNAME')).to.equal('x');
+    cy.get('[data-testid=username]').type(Cypress.env('LOGIN_USERNAME'));
```
Is searching by test id the recommended way of filling the forms? Or is there more user-oriented way, like in React testing library?

janybravo replied:
Its good practise to decouple element reference from code references class/id/css selection. By this code is more flexible to changes and is recomemended by eg. Cypress.

dblenkus commented on tsconfig.json:
```diff
@@ -14,7 +14,7 @@
         "isolatedModules": true,
         "noEmit": true,
-        "jsx": "react"
+        "jsx": "react",
```
There should be no trailing coma in JSON.

janybravo replied:
true

dblenkus commented on .env:
```diff
@@ -1,4 +1,5 @@
-REACT_APP_API_URL=https://localhost:8001/api
-REACT_APP_REST_AUTH_URL=https://localhost:8001/rest-auth
+REACT_APP_API_URL=http://localhost:3000/api
```
Is this configuration still required if the URL is the same?

janybravo replied:
They're as relevant as before. Will not change that in this PR, added todo to my notes for improvement.

zigasinko commented on cypress/support/index.js:
```diff
@@ -0,0 +1,15 @@
+    // For debugging purposes
+    // expect(Cypress.env('LOGIN_USERNAME')).to.equal('x');
+    cy.get('[data-testid=username]').type(Cypress.env('LOGIN_USERNAME'));
+    cy.get('[data-testid=password]').type(Cypress.env('LOGIN_PASSWORD'));
```
Add a `log=false` option so that password remains hidden in all logs / videos.

janybravo replied:
yes, done that

zigasinko commented on cypress/support/index.js:
```diff
@@ -0,0 +1,15 @@
+    cy.wait('@initRoute');
+    cy.get('[data-testid=open-login-modal]').click();
+    // For debugging purposes
+    // expect(Cypress.env('LOGIN_USERNAME')).to.equal('x');
```
Remove not used code.

janybravo replied:
utility comment for debugging environment variables to pass through, should remain

zigasinko commented on package.json:
```diff
@@ -38,16 +38,17 @@
         "build": "react-scripts build",
-        "test": "react-scripts test --env=jest-environment-jsdom-sixteen ",
+        "test": "react-scripts test --env=jest-environment-jsdom-sixteen",
+        "cy": "cross-env CYPRESS_LOGIN_USERNAME=${USERNAME:-ci-genialis} CYPRESS_LOGIN_PASSWORD=${PASS:-wrong123} cypress",
```
Duplicated logic for setting environment variables, extract it into a new script.

janybravo replied:
simplified

zigasinko commented on package.json:
```diff
@@ -38,16 +38,17 @@
-        "test": "react-scripts test --env=jest-environment-jsdom-sixteen ",
+        "test": "react-scripts test --env=jest-environment-jsdom-sixteen",
+        "cy": "cross-env CYPRESS_LOGIN_USERNAME=${USERNAME:-ci-genialis} CYPRESS_LOGIN_PASSWORD=${PASS:-wrong123} cypress",
+        "cy-run": "BROWSER=none cypress run",
```
`BROWSER=none` should apply to all cypress stuff, move setting this variable to a environment variables script specified above.

janybravo replied:
done

zigasinko commented on README.md:
```diff
@@ -47,6 +43,10 @@ Prettier and ESLint checks with enabled automatic fix combined.
 Launches the test runner in the interactive watch mode.<br />
 See the section about [running tests](https://facebook.github.io/create-react-app/docs/running-tests) for more information.
 
+### `LOGIN_USERNAME=myuser LOGIN_USERNAME=mysecret npm run cy open`
```
Add an explanation that this type of command only works on linux systems. Windows developers should use linux subsystem. Steps to make it work:
- login as root,
- npm install,
- install cypress: npx cypress install,
- I also had a problem with running create-react-app in WSL and had to add a PATH to the System32 mounted folder: [https://github.com/facebook/create-react-app/issues/7251](https://github.com/facebook/create-react-app/issues/7251),
- install Linux dependencies, instructions here: https://docs.cypress.io/guides/guides/continuous-integration.html#Advanced-setup.

zigasinko replied:
We found a better option that is platform agnostic! Remove cross-env from npm script:
`"test:e2e": "start-server-and-test start 3000 cy-run",`
fix readme instructions:
`npx cross-env CYPRESS_LOGIN_USERNAME=username CYPRESS_LOGIN_PASSWORD=pass npm run test:e2e`
and add a default environment variable `"LOGIN_USERNAME": "ci-genialis"` to `cypress.json`.

janybravo replied:
Removed variables alltogether from scripts as not much value it brings. See README.md for instructions.

zigasinko replied:
Unfortunately cross-env can't be avoided for setting env variables on Windows. Lean npm scripts working for me:
```
"cypress:run": "cypress run",
"test:e2e": "cross-env BROWSER=none start-server-and-test start 3000 cypress:run",
```

janybravo commented on package.json:
```diff
@@ -83,11 +84,14 @@
         "@types/redux-mock-store": "^1.0.2",
         "@types/styled-components": "^5.1.0",
         "browser-sync": "^2.26.7",
+        "cross-env": "^7.0.3",
```
@zigasinko  please verify on windows if still needed

zigasinko commented on cypress/support/index.js:
```diff
@@ -9,7 +9,7 @@ Cypress.Commands.add('login', () => {
     // expect(Cypress.env('LOGIN_USERNAME')).to.equal('x');
-    cy.get('[data-testid=username]').type(Cypress.env('LOGIN_USERNAME'));
-    cy.get('[data-testid=password]').type(Cypress.env('LOGIN_PASSWORD'));
+    cy.get('[data-testid=username]').type(Cypress.env('LOGIN_USERNAME'), { log: false });
```
I think logging username is fine. Comment was intended only for password.

zigasinko commented on package.json:
```diff
@@ -41,16 +41,16 @@
-        "cy": "cross-env CYPRESS_LOGIN_USERNAME=${USERNAME:-ci-genialis} CYPRESS_LOGIN_PASSWORD=${PASS:-wrong123} cypress",
-        "cy-run": "BROWSER=none cypress run",
-        "test:e2e": "cross-env CYPRESS_LOGIN_USERNAME=${LOGIN_USERNAME:-ci-genialis} CYPRESS_LOGIN_PASSWORD=${LOGIN_PASSWORD:-wrong123} BROWSER=none start-server-and-test start 3000 cy-run",
+        "cy": "cypress",
```
Do we need an alias for cypress? I'd remove and use `cypress` in scripts.

janybravo replied:
yes we do due to `start-server-and-test` within` test:e2e `

zigasinko commented on README.md:
````diff
@@ -43,9 +43,23 @@ Prettier and ESLint checks with enabled automatic fix combined.
+For development use UI variant (assuming locally running web site):
+
+```
+CYPRESS_LOGIN_USERNAME=myuser CYPRESS_LOGIN_USERNAME=mysecret npm run cy open
````
To set .env variables on Windows, we need to use "npx cross-env CYPRESS_LOGIN_USERNAME...".

janybravo replied:
fixed that

janybravo committed:
Fix linting scripts

Jenkins build fail when tsc/prettier/eslint returns error except for
JSON files.

janybravo committed:
Add e2e testing config and login e2e test spec

Change development proxy and update Jenkins CI pipeline script.

janybravo committed:
Add Expression time Courses graph e2e test spec

zigasinko commented on cypress/integration/expressionTimeCourses.spec.tsx:
```diff
@@ -0,0 +1,32 @@
+        cy.get(selectors.genesInput).clear().type('f');
+        cy.get(selectors.option).eq(2).click();
+
+        cy.get(selectors.expressionsGraph).matchImageSnapshot('expressions-graph', {
```
Snaptshot doesn't match on windows:
![expressions-graph diff](https://user-images.githubusercontent.com/26134756/105475047-3c877d80-5c9f-11eb-926d-e31355540530.png)

While fixing this, consider selecting genes via clipboard paste: `aplA, eif3L, fnkF_ps`, because cypress in open mode times out while selecting genes from autocomplete.

janybravo replied:
added bigger threshold, hopefully will resolve platform differencies

zigasinko commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 .test
 /coverage
+cypress/videos
+cypress/screenshots
```
Add `__diff_output__` (inside snapshots folder, if snapshot doesn't match) to .gitignore.

janybravo replied:
done that

zigasinko commented on cypress/integration/expressionTimeCourses.spec.tsx:
```diff
@@ -0,0 +1,53 @@
+
+            cy.get(selectors.genesInput).trigger('focus');
+            cy.get(selectors.genesInput).paste({ pastePayload: genes });
+            cy.wait(2e3);
```
Change that it waits for finished request to `${apiUrl}/kb/feature/search` or `GeneChip` elements are put into DOM.

janybravo committed:
fixup! Add Expression time Courses graph e2e test spec

janybravo committed:
fixup! Add Expression time Courses graph e2e test spec

janybravo committed:
Add complementary e2e test with Playwright/Jest
